arlas-core/src/main/java/io/arlas/server/impl/cache/HazelcastCacheManager.java

```diff
@@ -0,0 +1,126 @@
+ * under the License.
+ */
+
+package io.arlas.server.impl.cache;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.HazelcastInstanceNotActiveException;
+import io.arlas.server.app.ArlasServerConfiguration;
+import io.arlas.server.managers.CacheManager;
+import io.arlas.server.model.CollectionReference;
+import io.arlas.server.model.response.ElasticType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This cache holds a replicated map (named 'collections') for storing the collection references
+ * and one replicated map per collection (named '<collection name>') for storing the elastic types.
+ */
+public class HazelcastCacheManager implements CacheManager {
+    Logger LOGGER = LoggerFactory.getLogger(HazelcastCacheManager.class);
+    final private Config hzConfig;
+    final private ArlasServerConfiguration arlasConfig;
+    private HazelcastInstance instance;
+
+    public HazelcastCacheManager(ArlasServerConfiguration configuration) {
+        this.arlasConfig = configuration;
+        this.hzConfig = new Config();
+        hzConfig.setProperty( "hazelcast.phone.home.enabled", "false" );
+        // no need to expose the following env variable as a server configuration as it is set by Arlas Cloud if needed
+        String dns = System.getenv("ARLAS_CLOUD_SERVER_DNS");
+        if (dns != null) {
+            this.hzConfig.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+            this.hzConfig.getNetworkConfig().getJoin().getKubernetesConfig().setEnabled(true)
+                    .setProperty("service-dns", dns)
+                    .setProperty("service-dns-timeout", "60");
+        }
+        init();
+    }
+
+    private void init() {
+        LOGGER.info("Starting Hazelcast member");
+        this.instance = Hazelcast.newHazelcastInstance(this.hzConfig);
+    }
+
+    @Override
+    protected void finalize() throws Throwable {
+        super.finalize();
+        LOGGER.info("Stopping Hazelcast member");
+        instance.shutdown();
+    }
+
+    @Override
+    public CollectionReference getCollectionReference(String ref) {
+        try {
+            return (CollectionReference) this.instance.getReplicatedMap("collections").get(ref);
+        } catch (HazelcastInstanceNotActiveException e) { // recover from unexpected shutdown
+            init();
+            return (CollectionReference) this.instance.getReplicatedMap("collections").get(ref);
+        }
+    }
+
+    @Override
+    public void putCollectionReference(String ref, CollectionReference col) {
+        try {
+            // clear related ElasticTypes if cached
+            this.instance.getReplicatedMap(ref).clear();
+            this.instance.getReplicatedMap("collections").put(ref, col, arlasConfig.arlasCacheTimeout, TimeUnit.SECONDS);
```

Comment:
Does the `getReplicatedMap` method creates the map if it doesn't exist ?

Reply:
Yes. From javadoc: "Creates or returns the replicated map instance with the specified name."

Reply:
LGTM then